sdk/eventhub/azure-eventhubs-aio-partitionmanager-blobstorage/azure/eventhub/aio/partitionmanager/blobstorage/blobstoragepm.py

```diff
@@ -0,0 +1,125 @@
+                    "consumer_group_name": consumer_group_name,
+                    "partition_id": b.name,
+                    "owner_id": metadata["owner_id"],
+                    "etag": b.etag,
+                    "last_modified_time": b.last_modified.timestamp() if b.last_modified else None
+                }
+                ownership.update(metadata)
+                self._cached_ownership_dict[b.name] = ownership
+        return self._cached_ownership_dict.values()
+
+    async def claim_ownership(self, ownership_list: Iterable[Dict[str, Any]]) -> Iterable[Dict[str, Any]]:
+        result = []
+        for ownership in ownership_list:  # TODO: claiming concurrently
+            partition_id = ownership["partition_id"]
+            eventhub_name = ownership["eventhub_name"]
+            consumer_group_name = ownership["consumer_group_name"]
+            owner_id = ownership["owner_id"]
+
+            async with self._cached_ownership_locks[partition_id]:
+                metadata = {"owner_id": ownership["owner_id"]}
+                if "offset" in ownership:
+                    metadata["offset"] = ownership["offset"]
+                if "sequence_number" in ownership:
+                    metadata["sequence_number"] = ownership["sequence_number"]
+
+                etag = ownership.get("etag")
+                if etag:
+                    etag_match = {"if_match": '"'+etag+'"'}
+                else:
+                    etag_match = {"if_none_match": '"*"'}
+                try:
+                    blob_client = await self._container_client.upload_blob(
+                        name=partition_id, data=UPLOAD_DATA, overwrite=True, metadata=metadata, **etag_match
+                    )
+                    uploaded_blob_properties = await blob_client.get_blob_properties()
+                    ownership["etag"] = uploaded_blob_properties.etag
+                    ownership["last_modified_time"] = uploaded_blob_properties.last_modified.timestamp()
+                    self._cached_ownership_dict[partition_id] = ownership
+                except (ResourceModifiedError, ResourceExistsError):
+                    logger.info(
+                        "EventProcessor instance %r of eventhub %r consumer group %r lost ownership to partition %r",
+                        owner_id, eventhub_name, consumer_group_name, partition_id)
+                except Exception as err:
+                    logger.exception("An exception occurred when EventProcessor instance %r claim_ownership for "
+                                     "eventhub %r consumer group %r partition %r. The ownership is now lost",
+                                     owner_id, eventhub_name, consumer_group_name, partition_id, exc_info=err)
+                else:
+                    result.append(ownership)
+        return result
+
+    async def update_checkpoint(self, eventhub_name, consumer_group_name, partition_id, owner_id,
```

Comment:
There is code duplication/very similar code between `claim_ownership` and `update_checkpoint`. Refactor/share?

Reply:
The short exception handling logic looks similar but not the same. I would prefer keep it as is. Refactoring won't make it look better I think.